File: 35.py
# -*- coding:utf-8 -*-
class RandomListNode:
    def __init__(self, x):
        self.label = x
        self.next = None
        self.random = None
class Solution:
    # 返回 RandomListNode
    def Clone(self, pHead:RandomListNode):
        hashMap={}
        pNode = pHead
        while(pNode!=None):
            hashMap[pNode] = RandomListNode(pNode.label)
            pNode = pNode.next
        for key in hashMap:
            hashMap[key].next = hashMap[key.next]
            hashMap[key].random = hashMap[key.random]
        return hashMap[pHead]
# write code here

File: 35.cpp
//
// Created by zb198 on 2019/9/22.
//
#include <map>
using namespace std;
struct RandomListNode {
    int label;
    struct RandomListNode *next, *random;
    RandomListNode(int x) :
            label(x), next(nullptr), random(nullptr) {
    }
};

class Solution {
public:
    RandomListNode* Clone(RandomListNode* pHead)
    {
        map<RandomListNode&,RandomListNode&> hashMap;
        RandomListNode* pNode = pHead;
        while(pNode->next!= nullptr){
            hashMap[pNode] = RandomListNode(pNode->label);
            pNode = pNode->next;
        }
        for(auto Iter = hashMap.begin();Iter!=hashMap.end();Iter++){
            Iter->second.next = hashMap[Iter->first.next];
            Iter->second.random = hashMap[Iter->first.random];
        }
        return hashMap[pHead];
    }
};
